packages/utils-debugging-protocol-common/src/debugging-protocol-connector.ts

```diff
@@ -311,214 +258,30 @@ export class Connector implements IConnector {
-    private async createResponse(cdpResponse: Crdp.Network.ResponseReceivedEvent, element: IAsyncHTMLElement | null): Promise<Response> {
-        const resourceUrl: string = cdpResponse.response.url;
-        const hops: Array<string> = this._redirects.calculate(resourceUrl);
-        const resourceHeaders = normalizeHeaders(cdpResponse.response.headers);
-        const { content, rawContent, rawResponse } = await this.getResponseBody(cdpResponse);
-        const response: Response = {
-            body: {
-                content,
-                rawContent: rawContent!,
-                rawResponse: rawResponse as () => Promise<Buffer>
-            },
-            charset: null!,
-            headers: resourceHeaders!,
-            hops,
-            mediaType: null!,
-            statusCode: cdpResponse.response.status,
-            url: resourceUrl
-        };
-        const { charset, mediaType } = getContentTypeData(element, resourceUrl, response.headers, response.body.rawContent);
-
-        response.mediaType = mediaType!;
-        response.charset = charset!;
-
-        return response;
+        await this._server.emitAsync(eventName, event);
     }
 
-    /** Event handler fired when HTTP response is available and DOM loaded. */
-    private async onResponseReceived(params: Crdp.Network.ResponseReceivedEvent) {
-        const resourceUrl: string = params.response.url;
-        const hops: Array<string> = this._redirects.calculate(resourceUrl);
+    private async emitFetchEnd(requestResponse: RequestResponse) {
+        const resourceUrl: string = requestResponse.finalUrl;
+        const hops: Array<string> = requestResponse.hops;
         const originalUrl: string = hops[0] || resourceUrl;
 
         let element = null;
         let eventName: string = 'fetch::end';
         const isTarget: boolean = this._href === originalUrl;
 
         if (!isTarget) {
-            // DOM is not ready so we queue up the event for later
-            if (!this._dom) {
-                this._pendingResponseReceived.push(this.onResponseReceived.bind(this, params));
-
-                return;
-            }
-
             try {
-                element = await this.getElementFromRequest(params.requestId);
+                element = await this.getElementFromRequest(requestResponse.requestId);
             } catch (e) {
-                debug(`Error finding element for request ${params.requestId}. element will be null`);
+                debug(`Error finding element for request ${requestResponse.requestId}. element will be null`);
             }
         }
 
-        const response: Response = await this.createResponse(params, element);
+        const response: Response = requestResponse.getResponse(element);
 
         const request: Request = {
-            headers: params.response.requestHeaders as HttpHeaders,
+            headers: normalizeHeaders(requestResponse.responseReceived!.response.requestHeaders) as HttpHeaders,
```

Comment:
Nit: similar to my earlier comment about `this._dom`, consider passing `responseReceived` down as a parameter from the calling functions that checked to ensure it existed to make the fact that this is not `undefined` clearer (and understandable by the compiler).